File: go/inspection.go
package agentledger

type UnresolvedAttempt struct {
	RunID            string `json:"run_id"`
	LaneID           string `json:"lane_id"`
	TurnID           string `json:"turn_id"`
	ActionID         string `json:"action_id"`
	ActionType       string `json:"action_type"`
	AttemptID        string `json:"attempt_id"`
	AttemptNo        int    `json:"attempt_no"`
	RequestedEventID string `json:"requested_event_id"`
}

type LinkedCheckpoint struct {
	Event          StoredEvent    `json:"event"`
	CheckpointID   string         `json:"checkpoint_id,omitempty"`
	Profile        string         `json:"profile,omitempty"`
	ProfileVersion string         `json:"profile_version,omitempty"`
	Metadata       map[string]any `json:"metadata"`
}

type RunInspection struct {
	RunID              string              `json:"run_id"`
	TerminalEvents     []StoredEvent       `json:"terminal_events"`
	LinkedCheckpoints  []LinkedCheckpoint  `json:"linked_checkpoints"`
	UnresolvedAttempts []UnresolvedAttempt `json:"unresolved_attempts"`
}

func InspectRun(view RunView) RunInspection {
	result := RunInspection{RunID: view.RunID}
	for _, event := range view.Events {
		switch event.EventType {
		case EventTypeRunCompleted, EventTypeRunFailed, EventTypeRunCancelled:
			if event.SubjectID == view.RunID {
				result.TerminalEvents = append(result.TerminalEvents, event)
			}
		case EventTypeLaneFrameworkCheckpointLinked:
			metadata, _ := event.Payload["metadata"].(map[string]any)
			if metadata == nil {
				metadata = map[string]any{}
			}
			checkpointID, _ := event.Payload["checkpoint_id"].(string)
			profile, _ := event.Payload["profile"].(string)
			profileVersion, _ := event.Payload["profile_version"].(string)
			result.LinkedCheckpoints = append(result.LinkedCheckpoints, LinkedCheckpoint{
				Event: event, CheckpointID: checkpointID, Profile: profile,
				ProfileVersion: profileVersion, Metadata: metadata,
			})
		}
	}
	result.UnresolvedAttempts = unresolvedAttempts(view)
	return result
}

func unresolvedAttempts(view RunView) []UnresolvedAttempt {
	lanes := make(map[string]Lane, len(view.Lanes))
	for _, lane := range view.Lanes {
		lanes[lane.ID] = lane
	}
	turns := make(map[string]Turn, len(view.Turns))
	for _, turn := range view.Turns {
		turns[turn.ID] = turn
	}
	actions := make(map[string]Action, len(view.Actions))
	for _, action := range view.Actions {
		actions[action.ID] = action
	}
	attempts := make(map[string]Attempt, len(view.Attempts))
	for _, attempt := range view.Attempts {
		attempts[attempt.ID] = attempt
	}
	open := make(map[string]UnresolvedAttempt)
	order := make([]string, 0)
	ordered := make(map[string]struct{})
	for _, event := range view.Events {
		if event.SubjectKind() != "attempt" {
			continue
		}
		attempt, ok := attempts[event.SubjectID]
		if !ok {
			continue
		}
		action, ok := actions[attempt.ActionID]
		if !ok {
			continue
		}
		turn, ok := turns[action.TurnID]
		if !ok {
			continue
		}
		lane, ok := lanes[turn.LaneID]
		if !ok {
			continue
		}
		switch event.EventType {
		case EventTypeAttemptRequested:
			if _, exists := ordered[attempt.ID]; !exists {
				order = append(order, attempt.ID)
				ordered[attempt.ID] = struct{}{}
			}
			open[attempt.ID] = UnresolvedAttempt{
				RunID: lane.RunID, LaneID: lane.ID, TurnID: turn.ID,
				ActionID: action.ID, ActionType: action.Type,
				AttemptID: attempt.ID, AttemptNo: attempt.AttemptNo,
				RequestedEventID: event.ID,
			}
		case EventTypeAttemptCompleted, EventTypeAttemptFailed:
			delete(open, attempt.ID)
		}
	}
	result := make([]UnresolvedAttempt, 0, len(open))
	for _, attemptID := range order {
		if attempt, ok := open[attemptID]; ok {
			result = append(result, attempt)
		}
	}
	return result
}

File: typescript/packages/core/src/inspection.ts
import type {
  Action, Attempt, JsonValue, Lane, RunView, StoredEvent, Turn,
} from "./types.js";
import { EventType } from "./vocabulary.js";

export interface UnresolvedAttempt {
  run_id: string;
  lane_id: string;
  turn_id: string;
  action_id: string;
  action_type: string;
  attempt_id: string;
  attempt_no: number;
  requested_event_id: string;
}

export interface LinkedCheckpoint {
  event: StoredEvent;
  checkpoint_id?: string;
  profile?: string;
  profile_version?: string;
  metadata: { [key: string]: JsonValue };
}

export interface RunInspection {
  run_id: string;
  terminal_events: StoredEvent[];
  linked_checkpoints: LinkedCheckpoint[];
  unresolved_attempts: UnresolvedAttempt[];
}

export function inspectRun(view: RunView): RunInspection {
  const terminalTypes = new Set<string>([
    EventType.RUN_COMPLETED,
    EventType.RUN_FAILED,
    EventType.RUN_CANCELLED,
  ]);
  const terminalEvents = view.events.filter(
    (event) => terminalTypes.has(event.event_type) && event.subject_id === view.run_id,
  );
  const linkedCheckpoints = view.events
    .filter((event) => event.event_type === EventType.LANE_FRAMEWORK_CHECKPOINT_LINKED)
    .map((event): LinkedCheckpoint => {
      const checkpointId = event.payload.checkpoint_id;
      const profile = event.payload.profile;
      const profileVersion = event.payload.profile_version;
      const metadata = event.payload.metadata;
      return {
        event,
        ...(typeof checkpointId === "string" ? { checkpoint_id: checkpointId } : {}),
        ...(typeof profile === "string" ? { profile } : {}),
        ...(typeof profileVersion === "string" ? { profile_version: profileVersion } : {}),
        metadata: isJsonObject(metadata) ? metadata : {},
      };
    });
  return structuredClone({
    run_id: view.run_id,
    terminal_events: terminalEvents,
    linked_checkpoints: linkedCheckpoints,
    unresolved_attempts: unresolvedAttempts(view),
  });
}

function unresolvedAttempts(view: RunView): UnresolvedAttempt[] {
  const lanes = new Map<string, Lane>(view.lanes.map((lane) => [lane.id, lane]));
  const turns = new Map<string, Turn>(view.turns.map((turn) => [turn.id, turn]));
  const actions = new Map<string, Action>(view.actions.map((action) => [action.id, action]));
  const attempts = new Map<string, Attempt>(
    view.attempts.map((attempt) => [attempt.id, attempt]),
  );
  const open = new Map<string, UnresolvedAttempt>();
  for (const event of view.events) {
    if (!event.event_type.startsWith("attempt.")) continue;
    const attempt = attempts.get(event.subject_id);
    const action = attempt === undefined ? undefined : actions.get(attempt.action_id);
    const turn = action === undefined ? undefined : turns.get(action.turn_id);
    const lane = turn === undefined ? undefined : lanes.get(turn.lane_id);
    if (attempt === undefined || action === undefined || turn === undefined || lane === undefined) {
      continue;
    }
    if (event.event_type === EventType.ATTEMPT_REQUESTED) {
      open.set(attempt.id, {
        run_id: lane.run_id,
        lane_id: lane.id,
        turn_id: turn.id,
        action_id: action.id,
        action_type: action.type,
        attempt_id: attempt.id,
        attempt_no: attempt.attempt_no,
        requested_event_id: event.id,
      });
    } else if (
      event.event_type === EventType.ATTEMPT_COMPLETED
      || event.event_type === EventType.ATTEMPT_FAILED
    ) {
      open.delete(attempt.id);
    }
  }
  return [...open.values()];
}

function isJsonObject(value: JsonValue | undefined): value is { [key: string]: JsonValue } {
  return value !== null && typeof value === "object" && !Array.isArray(value);
}
